src/main/java/shop/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+        CartItem existingCartItem = cartItems.stream()
+                .filter(cartItem -> cartItem.getBook().getId().equals(requestDto.getBookId()))
+                .findFirst()
+                .orElse(null);
+        if (existingCartItem != null) {
+            existingCartItem.setQuantity(requestDto.getQuantity() + existingCartItem.getQuantity());
+        } else {
+            CartItem cartItem = new CartItem();
+            cartItem.setShoppingCart(shoppingCart);
+            cartItem.setQuantity(requestDto.getQuantity());
+            cartItem.setBook(bookRepository.findById(requestDto.getBookId())
+                    .orElseThrow(() -> new EntityNotFoundException("Can`t get book by id: "
+                            + requestDto.getBookId())));
+            cartItems.add(cartItem);
+            cartItemRepository.save(cartItem);
+        }
+        shoppingCart.setCartItems(cartItems);
+        shoppingCart = shoppingCartRepository.save(shoppingCart);
+        ShoppingCartResponseDto shoppingCartResponseDto = shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCart);
+        shoppingCartResponseDto.setCartItems(cartItemMapper.toCartItemResponseDtos(cartItems));
+        return shoppingCartResponseDto;
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartResponseDto updateQuantityBook(Long cartItemId,
+                                                  UpdateQuantityBookRequestDto requestDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can`t get cart by id: " + cartItemId));
+        cartItem.setQuantity(cartItem.getQuantity() + requestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        ShoppingCart shoppingCart = cartItem.getShoppingCart();
+        ShoppingCartResponseDto shoppingCartResponseDto = shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCart);
+        shoppingCartResponseDto.setCartItems(
+                cartItemMapper.toCartItemResponseDtos(shoppingCart.getCartItems()));
+        return shoppingCartResponseDto;
+    }
+
+    @Override
+    public void removeBookFromShoppingCart(Authentication authentication, Long cartItemId) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't get cart item by id: " + cartItemId));
+        ShoppingCart shoppingCart = cartItem.getShoppingCart();
+        if (cartItem.getQuantity() <= 1) {
+            shoppingCart.removeItemFromCart(cartItem);
+            cartItemRepository.delete(cartItem);
+            if (shoppingCart.getCartItems().isEmpty()) {
+                shoppingCart.setCartItems(new HashSet<>());
+            }
+        } else {
+            cartItem.setQuantity(cartItem.getQuantity() - 1);
+            cartItemRepository.save(cartItem);
+        }
```

Comment:
Let's remove this logic, this method involves completely removing the product from the cart. You have an update method to change the quantity. For a better understanding, open the website of any store. What requests are sent to the API under the hood of each button? This method will be called when the user clicks the remove button on the corresponding item in the cart.